Change how prefix and variable tags interact with each other

kidder opened the pull request:
## Proposed changes

Change the behavior of  `db::add_prefix<Prefix, Tag, Args...>` when `Tag` is a `::Tags::Variables<TagsList>` (with `TagsList = tmpl::list<Tag0,...,TagN>`) from producing a new tag `Prefix<::Tags::Variables<PrefixedTagsList, Args...>` with `PrefixedTagsList` = `tmpl::list<Prefix<Tag0, Args...>, ...,Prefix<TagN, Args...>>` to producing a new tag `::Tags::Variables<PrefixedTagsList, Args...>`.

I am making this change because it simplifies the code in `DataStructures/DataBox/PrefixHelpers.hpp` and because I found the tag `deriv<Variables<deriv<Var0>,...,deriv<VarN>>` to be potentially confusing as I read it to hold second derivatives...   In addition, there are several prefix tags that set `type = typename Tag::type` which is incorrect if `Tag` is a `::Tags::Variables`.

In cases where `db::add_prefix<Prefix, Tag, Args...>`  was never called on a `::Tags::Variables`, change to `Prefix<Tag, Args...>`.

In cases where `db::add_prefix<Prefix, Tag, Args...>` was always called on a
`::Tags::Variables`, rename the type alias to`db::variables_tag_with_tags_prefixed_by<Prefix, Tag, Args...>`

Remove `db::get_variables_tags_list` and `db::variables_tag_with_tags_list`
as they no longer need to handle a `::Tag::Variables` with both an outer and inner prefixes.

In addition, there are a few cleaning commits, and a couple of added `static_assert`s based on errors triggered when making these changes.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

@wthrowe there are two remaining uses of `db::add_tag_prefix` in `src/Time/Actions/RecordTimeStepperData.hpp` and `tests/Unit/Time/Actions/Test_SelfStartActions.cpp` because tests don't use a `Variables` as the `type` of the tag that is being wrapped.  Is this intentional, or can I update the tests to use a `Variables`?

@nilsleiffischer there are still many remaining uses of `db::add_tag_prefix` in the linear solver code, again because tests don't use a `Variables` as the `type` of the tag that is being wrapped.  Is this intentional?

kidder committed:
Remove inheritance from db::PrefixTag for compute tags

kidder committed:
Remove name function for tags that can be auto-generated

This required changing Tag::name to db::tag_name<Tag> in a few places.

kidder committed:
Move Tags::TempTensor and its aliases to their own file

Created DataStructures/Tags/TempTensor.hpp

kidder committed:
Add more test tags for testing DataStructures

Rename namespace VariablesTestTags_detail -> TestHelpers::Tags
Rename scalar -> Scalar and tensor -> Vector
Add DataVector as default for VectorType for these tags.

kidder committed:
Add static_asserts for invalid math/copy/move of Prefixed Variables

Some prefix tags have a different `type` than the `type` of the tag they wrap.
In such cases, the math/copy/move operations involving two Variables types
with the same unwrapped tags would either:  silently do the wrong thing, trigger
a mismatched size ASSERT in PointerVector, or segfault, depending upon the
number_of_grid_points and number_of_independent_components in each Variables.

kidder committed:
Add static assertion checking the tags of a Variables are tensors

kidder committed:
Replace db::get_variables_tags_list<T> with T::tags_list in appropiate places

This is a first step towards simplifying how variables tags and prefix tags interact.

kidder committed:
Change how prefix tags interact with ::Tags::Variables

Change the behavior of  db::add_prefix<Prefix, Tag, Args...> when Tag is a
::Tags::Variables<TagsList> (with TagsList = tmpl::list<Tag0,...,TagN>)
from producing a new tag Prefix<::Tags::Variables<PrefixedTagsList, Args...>
with PrefixedTagsList = tmpl::list<Prefix<Tag0, Args...>, ...,
Prefix<TagN, Args...>> to producing a new tag
::Tags::Variables<PrefixedTagsList, Args...>.

In cases where db::add_prefix<Prefix, Tag, Args...>  was never called on a
::Tags::Variables, change to Prefix<Tag, Args...>.

Remove db::get_variables_tags_list and db::variables_tag_with_tags_list
as they no longer need to handle a ::Tag::Variables with both an outer and
inner prefixes.

Clean up Test_PrefixHelpers.

kidder committed:
fix clang-tidy errors

kidder commented:
- clang-tidy errors are from the boost root-finder that we have not been able to ignore
- mac build is broken (see #2397)

wthrowe commented on src/DataStructures/Variables.hpp:
```diff
@@ -86,6 +86,8 @@ class Variables<tmpl::list<Tags...>> {
                 "You must provide at least one tag to the Variables "
                 "for type inference");
+  static_assert(
+      tmpl2::flat_all<tt::is_a_v<Tensor, typename Tags::type>...>::value);
```
`flat_all_v`

wthrowe commented on src/Evolution/Actions/ComputeTimeDerivative.hpp:
```diff
@@ -34,17 +34,17 @@ namespace Actions {
- * `db::add_tag_prefix<step_prefix, variables_tag>`, where `step_prefix` is
- * retrieved from the `Metavariables::temporal_id`. For evolution systems this
- * prefix is generally `Tags::dt`.
+ * `db::variables_tag_with_tags_prefixed_by<step_prefix, variables_tag>`, where
```
This comment is out-of-date.  The action doesn't do any prefixing of anything anymore.

wthrowe commented on src/Time/Tags.hpp:
```diff
@@ -25,6 +25,7 @@
 #include "Time/Time.hpp"
 #include "Time/TimeStepId.hpp"
 #include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/IsA.hpp"
```
You don't use this.

wthrowe reviewed (changes requested):
The use of the time-stepping code on non-Variables is intentional and I would prefer if it continued to work.

wthrowe commented:
Thinking about this more, what I really want is to ensure that we can integrate a stack-based data structure in the future.  We've sort of been using double and array for that and I was thinking of adding a new container type with tags, but maybe it should be a "Variables\<double>" specialization.  If that idea seems sane, I retract my objection to the time changes.

nilsvu commented on src/ParallelAlgorithms/LinearSolver/Tags.hpp:
```diff
@@ -218,11 +215,6 @@ struct OrthogonalizationHistory : db::PrefixTag, db::SimpleTag {
 struct KrylovSubspaceBasis : db::PrefixTag, db::SimpleTag {
-  static std::string name() noexcept {
-    // No "Linear" prefix since a Krylov subspace always refers to a linear
-    // operator
```
Could you keep this comment around? E.g. add this:

```cpp
// No "Linear" prefix is added to the tag's `name()` because a Krylov subspace always refers to a linear operator
```

nilsvu commented on src/DataStructures/Tags/TempTensor.hpp:
```diff
@@ -0,0 +1,114 @@
+// Distributed under the MIT License.
```
We don't have separate `Tags/` directories anywhere else in the code as far as I'm aware. What do you think about naming this file `DataStructures/TempTensorTags.hpp` instead?

kidder replied:
The plan is to move towards having separate `Tags/` directories with separate files per tag in order to reduce dependencies.  This is the first step...

nilsvu commented on tests/Unit/Helpers/DataStructures/TestTags.hpp:
```diff
@@ -3,49 +3,79 @@
-}  // namespace VariablesTestTags_detail
+
+template <class Tag>
+struct PrependIndex : ::db::PrefixTag, ::db::SimpleTag {
```
Template this class on `UpLo` and `Dim`?

kidder replied:
I just removed this and `ContractFirst` as I don't use them

nilsvu commented on tests/Unit/Helpers/DataStructures/TestTags.hpp:
```diff
@@ -3,49 +3,79 @@
+};
+
+template <class Tag>
+struct ContractFirst : ::db::PrefixTag, ::db::SimpleTag {
```
"Contract" implies another index to contract with, but the `type` just removes the first index. Name `RemoveFirstIndex` instead?

nilsvu commented on src/DataStructures/Variables.hpp:
```diff
@@ -288,6 +288,10 @@ class Variables<tmpl::list<Tags...>> {
   SPECTRE_ALWAYS_INLINE Variables& operator+=(
       const Variables<tmpl::list<WrappedTags...>>& rhs) noexcept {
+    static_assert(
+        (std::is_same_v<typename Tags::type, typename WrappedTags::type> &&
```
Can you replace `&&` -> `and` here and below?

nilsvu commented on src/DataStructures/DataBox/PrefixHelpers.hpp:
```diff
@@ -29,210 +29,100 @@ template <template <typename...> class Wrapper, typename TagList,
+/// tag with Prefix, which may have additional template arguments Args.
+template <template <typename...> class Prefix, typename VariablesTag,
+          typename... Args>
+using variables_tag_with_tags_prefixed_by =
```
Isn't this metafunction just `db::add_tag_prefix` with the restriction that it accepts nothing but a Variables tag? I think you can just delete it and we use `db::add_tag_prefix` everywhere.

nilsvu commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -56,9 +57,11 @@ struct InitializeSystem {
     using fields_tag = typename system::fields_tag;
     using linear_operator_applied_to_fields_tag =
-        db::add_tag_prefix<LinearSolver::Tags::OperatorAppliedTo, fields_tag>;
+        db::variables_tag_with_tags_prefixed_by<
```
You can leave these as-is, just using `db::add_tag_prefix` (same applies to many occurrences below)

(If your reasoning for this change is that you want to assert the tag is actually a variables tag: I think it's not the responsibility of the prefix utility metafunction to assert that. The class working with the tag, in this case `InitializeSystem`, should assert that what it receives is a variables tag if it expects one.)

nilsvu commented on src/NumericalAlgorithms/RootFinding/TOMS748.hpp:
```diff
@@ -54,9 +54,8 @@ double toms748(const Function& f, const double lower_bound,
-  // clang-tidy: internal boost warning, can't fix it.
-  auto result = boost::math::tools::toms748_solve(  // NOLINT
-      f, lower_bound, upper_bound, tol, max_iters);
+  auto result = boost::math::tools::toms748_solve(f, lower_bound, upper_bound,
```
This change seems unrelated to the rest of the commit. Better factor it into a separate commit if that's not too painful to do.

nilsvu commented on src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp:
```diff
@@ -164,8 +163,7 @@ struct Gmres {
    * - `orthogonalization_iteration_id_tag` =
-   * `db::add_tag_prefix<LinearSolver::Tags::Orthogonalization,
-   * LinearSolver::Tags::IterationId>`
+   * `LinearSolver::Tags::Orthogonalization< LinearSolver::Tags::IterationId>`
```
nitpick: delete space, same in line 214

nilsvu reviewed (changes requested):
The linear solver uses `db::add_tag_prefix` because it doesn't care if the tag holds a Variables or some vector type such as `DataVector` or `DenseVector`. It only requires the type supports addition, scalar multiplication and an inner product.

nilsdeppe commented:
The idea of a `Variables<double>` specialization is something that I would be happy to see. I've thought a bit about that in the past, since we could continue to get vector instructions through Blaze that way by using the static vector class. Adding a `Requires` specilization on Variables for the tag type is probably the way to go unless someone has a better idea. That way at least the memory layout is hidden from users.

wthrowe commented:
After looking into it, it looks like "Variables\<double>" (whether a specialization of Variables or not) is impossible to write without absurd tradeoffs.  We can't do Blaze math on Tensors because accessing the contents of multiple tensors as a big array violates C++'s stupid aliasing rules.  Allowing non-owning Tensors would work, but that has a memory overhead of more than a factor of 2.

I need to think a bit about how this affects what changes I'm OK with to the time stepping code.  It's nice to be able to integrate a few ODEs without having to go through the DataVector infrastructure (I've done it with our std::array math), but this now looks like it will be useless for any real runs.

wthrowe commented:
Actually, I have one more idea.  I need to flesh out the details.

kidder commented:
@wthrowe @nilsleiffischer I've pushed fixups.  In the interest of getting this merged sooner (as DataBox simplifications depend upon this) I've reverted the places I had replaced `add_tag_prefix` with `variables_tag_with_tags_prefixed_by` so they will work as before on either Variables or a non-Variables.  We can revisit this in the future as this does not affect the DataBox simplifications I am working on

nilsvu commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -42,7 +42,8 @@ namespace Actions {
  *   - `fields_tag`
- *   - `db::add_tag_prefix<::Tags::FixedSource, fields_tag>`
+ *   - `db::add_tag_prefix<::Tags::FixedSource,
+ * fields_tag>`
```
The revert commit left over a bunch of these formatting changes

nilsvu reviewed (dismissed):
Ok LGTM @kidder. Please don't wait for my re-approval if anything gets changed, since I'll be on vacation.

wthrowe commented:
Looks good.  Squash, etc.

kidder commented:
@wthrowe I've squashed, etc.

wthrowe reviewed (dismissed):
Clang-tidy failure is Boost internals.

nilsdeppe commented on src/DataStructures/Tags.hpp:
```diff
@@ -87,7 +83,8 @@ struct Multiplies : db::PrefixTag, db::SimpleTag {
   static std::string name() noexcept {
-    return "Multiplies(" + LhsTag::name() + ", " + RhsTag::name() + ")";
+    return "Multiplies(" + db::tag_name<LhsTag>() + ", " +
+           db::tag_name<RhsTag>() + ")";
```
Need an include for tagname

nilsdeppe reviewed (commented):
I'm happy with this. Just one missing include, the rest looks fine to me

kidder commented:
@nilsdeppe @wthrowe fixed the include

nilsdeppe reviewed (approved):
CI error is clang-tidy in boost internals
